File: GreedyStarAssistant/app/src/main/java/com/greedy/assistant/DynamicMarkov.java
package com.greedy.assistant;

import android.content.Context;
import android.content.SharedPreferences;
import org.json.JSONArray;
import org.json.JSONObject;
import java.io.InputStream;
import java.util.HashMap;
import java.util.Map;

public class DynamicMarkov {
    private Map<String, int[]> transition2 = new HashMap<>();
    private int[][] transition1 = new int[8][8];
    private SharedPreferences prefs;

    public DynamicMarkov(Context context) {
        prefs = context.getSharedPreferences("greedy_bot", Context.MODE_PRIVATE);
        loadBaseModel(context);
    }

    private void loadBaseModel(Context context) {
        try {
            InputStream is = context.getAssets().open("shared_data.json");
            int size = is.available();
            byte[] buffer = new byte[size];
            is.read(buffer);
            is.close();
            String jsonStr = new String(buffer, "UTF-8");
            JSONObject root = new JSONObject(jsonStr);

            // transitionCounts1
            JSONArray tc1 = root.getJSONArray("transitionCounts1");
            for (int i = 0; i < tc1.length() && i < 8; i++) {
                JSONArray row = tc1.getJSONArray(i);
                for (int j = 0; j < row.length() && j < 8; j++) {
                    transition1[i][j] = row.getInt(j);
                }
            }

            // transitionCounts2
            JSONObject tc2 = root.getJSONObject("transitionCounts2");
            java.util.Iterator<String> keys = tc2.keys();
            while (keys.hasNext()) {
                String key = keys.next();
                String[] parts = key.split(",");
                if (parts.length == 2) {
                    int a = Integer.parseInt(parts[0]);
                    int b = Integer.parseInt(parts[1]);
                    JSONArray arr = tc2.getJSONArray(key);
                    int[] counts = new int[8];
                    for (int i = 0; i < arr.length() && i < 8; i++) counts[i] = arr.getInt(i);
                    transition2.put(a + "," + b, counts);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public int predict(int prevPrev, int prev) {
        String key = prevPrev + "," + prev;
        int[] t2 = transition2.get(key);
        if (t2 != null && sum(t2) > 0) {
            return argMax(t2);
        }
        if (prev >= 0 && prev < 8) {
            return argMax(transition1[prev]);
        }
        return 0;
    }

    public void update(int prevPrev, int prev, int correct) {
        String key = prevPrev + "," + prev;
        int[] t2 = transition2.get(key);
        if (t2 == null) {
            t2 = new int[8];
            transition2.put(key, t2);
        }
        t2[correct]++;
        if (prev >= 0 && prev < 8) {
            transition1[prev][correct]++;
        }
        saveUpdates();
    }

    private int sum(int[] arr) {
        int s = 0;
        for (int v : arr) s += v;
        return s;
    }

    private int argMax(int[] arr) {
        int maxIdx = 0;
        for (int i = 1; i < arr.length; i++) if (arr[i] > arr[maxIdx]) maxIdx = i;
        return maxIdx;
    }

    private void saveUpdates() {
        // يمكن حفظ transition2 هنا إلى SharedPreferences إذا أردت الاحتفاظ بالتحديثات
    }
}

File: GreedyStarAssistant/app/src/main/java/com/greedy/assistant/DynamicMarkov.kt
package com.greedy.assistant

import android.content.Context
import kotlinx.coroutines.*
import org.json.JSONObject
import java.io.InputStream

class DynamicMarkov(private val context: Context) {
    private var transition2 = mutableMapOf<Pair<Int,Int>, IntArray>()
    private var transition1 = Array(8) { IntArray(8) }

    suspend fun loadModel() = withContext(Dispatchers.IO) {
        val json = context.assets.open("shared_data.json").bufferedReader().use { it.readText() }
        val root = JSONObject(json)
        // تحميل transitionCounts1, transitionCounts2 من JSON
        // (نفس الكود السابق لكن مع coroutines)
    }

    fun predict(prevPrev: Int, prev: Int): Int {
        val key = prevPrev to prev
        val t2 = transition2[key]
        if (t2 != null && t2.sum() > 0) return t2.indices.maxByOrNull { t2[it] } ?: 0
        if (prev in 0..7) return transition1[prev].indices.maxByOrNull { transition1[prev][it] } ?: 0
        return 0
    }

    fun update(prevPrev: Int, prev: Int, correct: Int) {
        val key = prevPrev to prev
        val t2 = transition2.getOrPut(key) { IntArray(8) }
        t2[correct]++
        transition1[prev][correct]++
    }
}